Improve Performance of Plug.Conn.Cookies.decode/1

isavita opened the pull request:
# Description
This pull request optimizes the Plug.Conn.Cookies.decode/1 function to improve its performance.

This PR addresses issue [#1227](https://github.com/elixir-plug/plug/issues/1227).

josevalim commented:
Hi @isavita! I have already pushed a fix. I compared it with yours and the current version in main is faster. Please double check you see the same result.

isavita commented:
@josevalim 
My solution has failing test
```bash
  1) test decode cookies (Plug.Conn.CookiesTest)
     test/plug/conn/cookies_test.exs:7
     Assertion with == failed
     code:  assert decode("key=#{whitespace}") == %{}
     left:  %{"key" => ""}
     right: %{}
     stacktrace:
       test/plug/conn/cookies_test.exs:31: anonymous fn/2 in Plug.Conn.CookiesTest."test decode cookies"/1
       (elixir 1.14.3) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/plug/conn/cookies_test.exs:24: (test)
```

Here is benchmark
```elixir
defmodule CookiesJose do
  def decode(cookie) when is_binary(cookie) do
    Map.new(decode_kv(cookie, []))
  end

  defp decode_kv("", acc), do: acc
  defp decode_kv(<<h, t::binary>>, acc) when h in [?\s, ?\t], do: decode_kv(t, acc)
  defp decode_kv(kv, acc) when is_binary(kv), do: decode_key(kv, "", acc)

  defp decode_key(<<h, t::binary>>, _key, acc) when h in [?\s, ?\t, ?\r, ?\n, ?\v, ?\f],
    do: skip_until_cc(t, acc)

  defp decode_key(<<?;, t::binary>>, _key, acc), do: decode_kv(t, acc)
  defp decode_key(<<?=, t::binary>>, "", acc), do: skip_until_cc(t, acc)
  defp decode_key(<<?=, t::binary>>, key, acc), do: decode_value(t, "", 0, key, acc)
  defp decode_key(<<h, t::binary>>, key, acc), do: decode_key(t, <<key::binary, h>>, acc)
  defp decode_key(<<>>, _key, acc), do: acc

  defp decode_value(<<?;, t::binary>>, value, spaces, key, acc),
    do: decode_kv(t, [{key, trim_spaces(value, spaces)} | acc])

  defp decode_value(<<?\s, t::binary>>, value, spaces, key, acc),
    do: decode_value(t, <<value::binary, ?\s>>, spaces + 1, key, acc)

  defp decode_value(<<h, t::binary>>, _value, _spaces, _key, acc)
       when h in [?\t, ?\r, ?\n, ?\v, ?\f],
       do: skip_until_cc(t, acc)

  defp decode_value(<<h, t::binary>>, value, _spaces, key, acc),
    do: decode_value(t, <<value::binary, h>>, 0, key, acc)

  defp decode_value(<<>>, value, spaces, key, acc),
    do: [{key, trim_spaces(value, spaces)} | acc]

  defp skip_until_cc(<<?;, t::binary>>, acc), do: decode_kv(t, acc)
  defp skip_until_cc(<<_, t::binary>>, acc), do: skip_until_cc(t, acc)
  defp skip_until_cc(<<>>, acc), do: acc

  defp trim_spaces(value, 0), do: value
  defp trim_spaces(value, spaces), do: binary_part(value, 0, byte_size(value) - spaces)
end

defmodule CookiesAlex do
  def decode(cookie) do
    cookie
    |> :binary.split(";", [:global])
    |> Enum.reduce(%{}, fn kv, acc ->
      case decode_kv(kv) do
        {k, v} -> Map.put(acc, k, v)
        false -> acc
      end
    end)
  end

  defp decode_kv(""), do: false
  defp decode_kv(kv) do
    kv
    |> String.trim()
    |> decode_key("")
  end

  defp decode_key("", _key), do: false
  defp decode_key(<<?=, _::binary>>, ""), do: false
  defp decode_key(<<?=, t::binary>>, key) do
    value = case t do
      "" -> ""
      _ -> String.trim(t)
    end
    {key, value}
  end
  defp decode_key(<<h, _::binary>>, _key) when h in [?\s, ?\t, ?\r, ?\n, ?\v, ?\f], do: false
  defp decode_key(<<h, t::binary>>, key), do: decode_key(t, <<key::binary, h>>)
end
defmodule CookiesOld do
  @moduledoc """
  Conveniences for encoding and decoding cookies.
  """

  @doc """
  Decodes the given cookies as given in either a request or response header.

  If a cookie is invalid, it is automatically discarded from the result.

  ## Examples

      iex> decode("key1=value1;key2=value2")
      %{"key1" => "value1", "key2" => "value2"}

  """
  def decode(cookie) do
    do_decode(:binary.split(cookie, ";", [:global]), %{})
  end

  defp do_decode([], acc), do: acc

  defp do_decode([h | t], acc) do
    case decode_kv(h) do
      {k, v} -> do_decode(t, Map.put(acc, k, v))
      false -> do_decode(t, acc)
    end
  end

  defp decode_kv(""), do: false
  defp decode_kv(<<h, t::binary>>) when h in [?\s, ?\t], do: decode_kv(t)
  defp decode_kv(kv), do: decode_key(kv, "")

  defp decode_key("", _key), do: false
  defp decode_key(<<?=, _::binary>>, ""), do: false
  defp decode_key(<<?=, t::binary>>, key), do: decode_value(t, "", key, "")
  defp decode_key(<<h, _::binary>>, _key) when h in [?\s, ?\t, ?\r, ?\n, ?\v, ?\f], do: false
  defp decode_key(<<h, t::binary>>, key), do: decode_key(t, <<key::binary, h>>)

  defp decode_value("", _spaces, key, value), do: {key, value}

  defp decode_value(<<?\s, t::binary>>, spaces, key, value),
    do: decode_value(t, <<spaces::binary, ?\s>>, key, value)

  defp decode_value(<<h, _::binary>>, _spaces, _key, _value) when h in [?\t, ?\r, ?\n, ?\v, ?\f],
    do: false

  defp decode_value(<<h, t::binary>>, spaces, key, value),
    do: decode_value(t, "", key, <<value::binary, spaces::binary, h>>)
end

defmodule Benchmark.CookiesDecode do
  def run(_) do
    cookie =
    "a=1; b=111; b_exp=1738257503023; c=july2023; b=fsasfd1d46277134810r0.11893443256905789; d=b9d4ad1d-94dc-7fd6-ae1d-305e63791ac6; e=[%22somestuff%22]; fff={%22volume%22:0.7}; anotherstuff=hh; aaa3=hh; bbb={}; uii=somemorerandomstuf; hhhh=morestuff; kkk=eyJhYiI6Im8xOCIsInBhIjoiQ003IiwiZXAiOnRydWUsImV6IjpmYWxzZSwicGwiOmZhbHNlLCJwcyI6IjE0V2VDNElVaTRJaU9hSlp2NnpfM0tSMTJDMjE0NUU3UVN4SjZ4RHJRNG8iLCJjbCI6ZmFsc2UsImNjIjoiZ2IiLCJyZWFsbSI6Ii8iLCJzZXMtZXhwIjoxNzEyMDYzNzkwMDAwLCJqd3QtZXhwIjoxNzc1MTM0ODkwMDAwLCJydGtuLWV4cCI6MTc3NTEzNDg5MDAwMCwidGtuLWV4cCI6MTcxODgxNzMwMjAwMCwicG4iOjM5fQ; _chartbeat2=.1706277135627.1718810705253.0000001011000111.Cbuhe6BNM-tUCX5KjaCce88iBsBqeL.2; _cccc=Bfhgdfgd_dfsafsKKK; some_nonve=D9Uze6r1L5Cb9iZE8BXHXfMU; some_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG8gbGRqa2Zza2xyZWpmYXNsa2RmamtzamFma2RsanNrYWxmamRza2xqZmtsZHNqZmtsZHNqZmtsYWpza2xkZmpzYWZkcyBNYWtlIHRoaXMgYXMgYmlnIGFzIHBvc3NpYmxlLkVsaXhpciBpcyBhIGR5bmFtaWMsIGZ1bmN0aW9uYWwgbGFuZ3VhZ2UgZGVzaWduZWQgZm9yIGJ1aWxkaW5nIHNjYWxhYmxlIGFuZCBtYWludGFpbmFibGUgYXBwbGljYXRpb25zLiBJdCBsZXZlcmFnZXMgdGhlIEVybGFuZyBWTSwga25vd24gZm9yIHJ1bm5pbmcgbG93LWxhdGVuY3ksIGRpc3RyaWJ1dGVkLCBhbmQgZmF1bHQtdG9sZXJhbnQgc3lzdGVtcy4gICAgIEVsaXhpciBpcyBhIGR5bmFtaWMsIGZ1bmN0aW9uYWwgbGFuZ3VhZ2UgZGVzaWduZWQgZm9yIGJ1aWxkaW5nIHNjYWxhYmxlIGFuZCBtYWludGFpbmFibGUgYXBwbGljYXRpb25zLiBJdCBsZXZlcmFnZXMgdGhlIEVybGFuZyBWTSwga25vd24gZm9yIHJ1bm5pbmcgbG93LWxhdGVuY3ksIGRpc3RyaWJ1dGVkLCBhbmQgZmF1bHQtdG9sZXJhbnQgc3lzdGVtcy5FbGl4aXIgaXMgYSBkeW5hbWljLCBmdW5jdGlvbmFsIGxhbmd1YWdlIGRlc2lnbmVkIGZvciBidWlsZGluZyBzY2FsYWJsZSBhbmQgbWFpbnRhaW5hYmxlIGFwcGxpY2F0aW9ucy4gSXQgbGV2ZXJhZ2VzIHRoZSBFcmxhbmcgVk0sIGtub3duIGZvciBydW5uaW5nIGxvdy1sYXRlbmN5LCBkaXN0cmlidXRlZCwgYW5kIGZhdWx0LXRvbGVyYW50IHN5c3RlbXMuIiwiaWF0IjoxNTE2MjM5MDIyfQ.g_440sF_734II0EyaWnOCI-nmM-AbQ41JPfl8gnv6Ow; kkktkn=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c; idid=Elixir%20is%20a%20dynamic%2C%20functional%20language%20designed%20for%20building%20scalable%20and%20maintainable%20applications.%20It%20leverages%20the%20Erlang%20VM%2C%20known%20for%20running%20low-latency%2C%20distributed%2C%20and%20fault-tolerant%20systems; _aaaa=random; TTT_fdsdfd=1; HHH_fdsfds=1"

    Benchee.run(
      %{
      "old: Plug.Conn.Cookies.decode/1" => fn ->
        CookiesOld.decode(cookie)
      end,
      "Plug.Conn.CookiesJose.decode/1" => fn ->
        CookiesJose.decode(cookie)
      end,
      "Plug.Conn.CookiesAlex.decode/1" => fn ->
        CookiesAlex.decode(cookie)
      end
      },
      time: 5,
      memory_time: 2
    )
  end
end
```
between 3 versions
```bash
Operating System: macOS
CPU Information: Apple M2 Pro
Number of Available Cores: 12
Available memory: 16 GB
Elixir 1.16.2
Erlang 26.2.5
JIT enabled: true

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 2 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 27 s

Benchmarking new2: Plug.Conn.Cookies.decode/1 ...
Benchmarking new: Plug.Conn.Cookies.decode/1 ...
Benchmarking old: Plug.Conn.Cookies.decode/1 ...
Calculating statistics...
Formatting results...

Name                                       ips        average  deviation         median         99th %
new: Plug.Conn.Cookies.decode/1       152.89 K        6.54 μs   ±168.62%        5.33 μs       23.71 μs
new2: Plug.Conn.Cookies.decode/1      115.56 K        8.65 μs    ±17.58%        8.08 μs       15.13 μs
old: Plug.Conn.Cookies.decode/1         3.58 K      279.37 μs    ±12.20%      277.58 μs      376.55 μs

Comparison:
new: Plug.Conn.Cookies.decode/1       152.89 K
new2: Plug.Conn.Cookies.decode/1      115.56 K - 1.32x slower +2.11 μs
old: Plug.Conn.Cookies.decode/1         3.58 K - 42.71x slower +272.83 μs

Memory usage statistics:

Name                                Memory usage
new: Plug.Conn.Cookies.decode/1         11.54 KB
new2: Plug.Conn.Cookies.decode/1        14.91 KB - 1.29x memory usage +3.37 KB
old: Plug.Conn.Cookies.decode/1        263.25 KB - 22.81x memory usage +251.71 KB
```

josevalim commented:
Thank you. I am getting different results on OTP 27. Yours is twice faster but uses twice more memory.

However, I think your solution is faster because it avoid traversing the values but you have to always traverse the values to discard them if they have of the space characters and discard them otherwise :(

isavita commented:
@josevalim Thank you! Any plans to do minor release something like `1.16.1`?

josevalim commented:
Done!
